### DISCUSSION TITLE
how to use msquic on my own udp socket

### DISCUSSION THREAD

**Discussion by MemeTao** [microsoft/msquic#3270] [Q&A]
I want use msquic on my app:

I cerate an udp socket, and provide 'send' 'recv'  to msquic, is there such an interface?

**Comment by nibanks** [comment]
Unfortunately, no. MsQuic doesn't support app-owned UDP sockets as we have to set a lot of custom socket options to make everything work. What are you trying to accomplish? Why do you need to own the socket?

**Reply by MemeTao** [reply]
thanks for reply.

my app is based on webrtc(there is only one udp port i can use)

webrtc data channel(sctp protocol) is very terrible to transfer data, so i want to use msquic to replace sctp.

**Reply by anrossi** [reply]
I'm not very familiar with the webrtc protocol, but does it give you a regular UDP socket? It sounds like it gives the user a way to read/write data, but the underlying protocol is already determined for the user (SCTP).

**Reply by MemeTao** [reply]
simplest way is use the unreliable mode of sctp, and then implement quic protocol on "send"/ "recv".

I want to abandoned sctp protocol totally, and implement quic protocol on regular udp socket.

**Reply by nibanks** [reply]
I still don't understand why you can't just move to using MsQuic API instead.

**Reply by chunlei-wu** [reply]
I guess the webrtc have their own peer connection channel, such as P2P channel or something else, so all the webrtc data should be send/received via the peer connection channel. If you want to send some data encapsulated in QUIC, you will hope to do the following:
1, start quic
2, connect
3, create stream
4, send the data to quic
5, quic return the udp datagram with quic frame
6, get the udp datagram
7, send the udp datagram via the peer connection channel
on the other hand, for the receive data:
1, receive the udp datagram with quic frame from the peer connection channel
2, call some API to make quic handle it as quic receive udp datagram from its socket
3, quic return the data to webrtc without the quic header

I also want to know how to handle the similar case, such as: I want to setup a tunnel protocol channel, I hope the user data will transfer by quic protocol via the tunnel channel, how to handle it?

**Reply by MemeTao** [reply]
That dude is right, I mean it too

**Reply by guest271314** [reply]
I'm working on something similar, trying to get a WebTransport server working in the browser over Direct Sockets `UDPSocket`. To do that, I think compiling the core functionality to WASM will work, reading from and writing to the UDP socket. Where exactly is the network socket created in MsQuic?

**Reply by guhetier** [reply]
@guest271314 If you are interested, you can look at CxPlatSocketCreateUdp and what happens from there in SocketCreateUdp, etc...
The code is somewhat convoluted, with different implementations for different platforms.

We might refactor this code soon, if we have a chance, but as of now, there is no straightforward way to bring your own socket.

**Reply by guest271314** [reply]
@guhetier 

> @guest271314 If you are interested, you can look at CxPlatSocketCreateUdp and what happens from there in SocketCreateUdp, etc... The code is somewhat convoluted, with different implementations for different platforms.
> 
> We might refactor this code soon, if we have a chance, but as of now, there is no straightforward way to bring your own socket.

Yeah, I don't think I can cite a repository off the top of my head that has a QUIC/WebTransport implementation where you can bring your own (UDP) socket. 

Targeting WASM/WASI for the TLS and protocol might also be useful, too.
